File: gift_shopping.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// to get user input
function getUserInput(productName) {
    return new Promise((resolve) => {
        rl.question(`Enter the quantity for ${productName}: `, (quantity) => {
            rl.question(`Is ${productName} wrapped as a gift? (yes/no): `, (isGift) => {
                resolve({ quantity: parseInt(quantity, 10), isGift: isGift.toLowerCase() === 'yes' });
            });
        });
    });
}

// to calculate total cost 
async function calculateTotalCost() {
    const priceA = 20;
    const priceB = 40;
    const priceC = 50;

    const flat10Discount = 10;
    const bulk5DiscountPercentage = 5;
    const bulk10DiscountPercentage = 10;
    const tiered50DiscountPercentage = 50;

    const giftWrapFee = 1;
    const shippingFeePerPackage = 5;
    const itemsPerPackage = 10;

    const productA = await getUserInput('Product A');
    const productB = await getUserInput('Product B');
    const productC = await getUserInput('Product C');

    const { quantity: quantityA, isGift: isGiftA } = productA;
    const { quantity: quantityB, isGift: isGiftB } = productB;
    const { quantity: quantityC, isGift: isGiftC } = productC;

    const totalQuantity = quantityA + quantityB + quantityC;
    const cartTotal = quantityA * priceA + quantityB * priceB + quantityC * priceC;

    let discountName = '';
    let discountAmount = 0;

    // for flat 10 discount
    if (cartTotal > 200) {
        discountName = "flat_10_discount";
        discountAmount = flat10Discount;
    }
    //  for bulk 10 discount
    else if (totalQuantity > 20) {
        discountName = "bulk_10_discount";
        discountAmount = (cartTotal * bulk10DiscountPercentage) / 100;
    }
    // for bulk 5 discount on any single product
    else if (quantityA > 10 || quantityB > 10 || quantityC > 10) {
        discountName = "bulk_5_discount";
        discountAmount = (quantityA > 10 ? quantityA * priceA : 0) * (bulk5DiscountPercentage / 100)
            + (quantityB > 10 ? quantityB * priceB : 0) * (bulk5DiscountPercentage / 100)
            + (quantityC > 10 ? quantityC * priceC : 0) * (bulk5DiscountPercentage / 100);
    }
    // for tiered 50 discount
    else if (totalQuantity > 30 && (quantityA > 15 || quantityB > 15 || quantityC > 15)) {
        discountName = "tiered_50_discount";
        discountAmount = (quantityA > 15 ? (quantityA - 15) * priceA : 0) * (tiered50DiscountPercentage / 100)
            + (quantityB > 15 ? (quantityB - 15) * priceB : 0) * (tiered50DiscountPercentage / 100)
            + (quantityC > 15 ? (quantityC - 15) * priceC : 0) * (tiered50DiscountPercentage / 100);
    }

    const subtotal = cartTotal - discountAmount;
    const shippingFee = Math.ceil(totalQuantity / itemsPerPackage) * shippingFeePerPackage;
    const total = subtotal + shippingFee + (isGiftA ? quantityA * giftWrapFee : 0) + (isGiftB ? quantityB * giftWrapFee : 0) + (isGiftC ? quantityC * giftWrapFee : 0);

    // Output details
    console.log("Product A - Quantity: " + quantityA + ", Total: $" + quantityA * priceA);
    console.log("Product B - Quantity: " + quantityB + ", Total: $" + quantityB * priceB);
    console.log("Product C - Quantity: " + quantityC + ", Total: $" + quantityC * priceC);
    console.log("Subtotal: $" + subtotal.toFixed(2));
    console.log("Discount Applied: " + discountName + ", Amount: $" + discountAmount.toFixed(2));
    console.log("Shipping Fee: $" + shippingFee);
    console.log("Gift Wrap Fee for Product A: $" + (isGiftA ? quantityA * giftWrapFee : 0));
    console.log("Gift Wrap Fee for Product B: $" + (isGiftB ? quantityB * giftWrapFee : 0));
    console.log("Gift Wrap Fee for Product C: $" + (isGiftC ? quantityC * giftWrapFee : 0));
    console.log("Total: $" + total.toFixed(2));

    rl.close();
}

//  to calculate total cost
calculateTotalCost();

File: gift_shopping.py
def calculate_discount(cart, product_quantities):
    total_quantity = sum(product_quantities.values())
    max_quantity = max(product_quantities.values())

    if total_quantity > 30 and max_quantity > 15:
        return "tiered_50_discount", sum(cart.values()) * 0.5
    elif total_quantity > 20:
        return "bulk_10_discount", sum(cart.values()) * 0.1
    elif max_quantity > 10:
        for product in product_quantities:
            if product_quantities[product] > 10:
                return "bulk_5_discount", cart[product] * 0.05
    elif sum(cart.values()) > 200:
        return "flat_10_discount", 10

    return None, 0

def calculate_total(cart, product_quantities, gift_wrap_fee, shipping_fee):
    discount_rule, discount_amount = calculate_discount(cart, product_quantities)
    subtotal = sum(cart.values())
    total_discounted_amount = subtotal - discount_amount
    total_gift_wrap_fee = sum(product_quantities.values()) * gift_wrap_fee
    total_shipping_fee = (sum(product_quantities.values()) // 10) * shipping_fee
    total = total_discounted_amount + total_gift_wrap_fee + total_shipping_fee

    return discount_rule, discount_amount, subtotal, total_gift_wrap_fee, total_shipping_fee, total

def display_receipt(cart, product_quantities, gift_wrap_fee, shipping_fee):
    print("Product Details:")
    for product in cart:
        print(f"{product}: Quantity: {product_quantities[product]}, Total: ${cart[product]}")

    discount_rule, discount_amount, subtotal, total_gift_wrap_fee, total_shipping_fee, total = calculate_total(
        cart, product_quantities, gift_wrap_fee, shipping_fee)

    print("\nSubtotal: ${}".format(subtotal))
    if discount_rule:
        print("Discount Applied: {} - ${}".format(discount_rule, discount_amount))
    print("Gift Wrap Fee: ${}".format(total_gift_wrap_fee))
    print("Shipping Fee: ${}".format(total_shipping_fee))
    print("\nTotal: ${}".format(total))

# Helper function for calculating product total
def calculate_product_total(quantity, price, gift_wrap_fee):
    return quantity * price + (gift_wrap_fee * quantity)

# Main function
def main():
    products = {"Product A": 20, "Product B": 40, "Product C": 50}
    gift_wrap_fee = 1
    shipping_fee = 5

    product_quantities = {}

    for product in products:
        quantity = int(input("Enter the quantity of {} you want to purchase: ".format(product)))
        is_gift_wrapped = input("Is {} gift wrapped? (yes/no): ".format(product)).lower() == "yes"
        
        total_amount = calculate_product_total(quantity, products[product], gift_wrap_fee) \
                        if is_gift_wrapped else calculate_product_total(quantity, products[product], 0)
        
        product_quantities[product] = quantity
        print("{}: Quantity: {}, Total Amount: ${}".format(product, quantity, total_amount))

    display_receipt(products, product_quantities, gift_wrap_fee, shipping_fee)

if __name__ == "__main__":
    main()
